### PR TITLE
LuaRocks 3.x

### PR DESCRIPTION
**Pull request by zwirec**

### CODE REVIEW

**Comment by Totktonada** [comment]
Usually we comment a patchset, but here the deadline is near, so I decided to
make changes myself. I have pushed the original branch to `luarocks-3.x-orig`
branch, so you can verify the changes.

I reverted some of changes, because failed to understood what is a problem and
how to check that it is resolved. I hope you'll answer here whether they are
actually needed and why; and we'll proceed with them (within this PR or
another ones depending on timing). See details below.

Below I have summarized changes I made for the patchset, its motivation and
checks I performed during the review.

Re first commit ('Fix compatibility with tarantool static build'):

* Squashed two commits re compatibility with tarantool: 'Fix compatibility
  with tarantool static build' and 'patch defaults for using with tarantool'.
* Updated the description of the commit re compatibility with tarantool:
  described all made changes, mention relevant commits from `tarantool-1.7`
  branch, listed all hardcoded.lua options that was added or ported from
  luarocks-2.
* Removed `localdir = require('fio').cwd()` in `env_for_config_file()`,
  because in set afterwards in `set_confdirs()` using `hardcoded.LOCALDIR`
  value. (@rosik suggested me that it possibly set here to be seen in
  hardcoded.lua, but it is not used in tarantool's hardcoded.lua file.)
* Fixed indentation in the 'rocks_provided' code block.
* Renamed cfg.inited to cfg.initialized. BTW, don't get why it is needed, but
  don't want to dig into that.
* Removed 'we don't need to workaround older luarocks' comment from the code
  (because it actually deletes the code which comments and becomes dangling),
  commented it in the commit message instead.

Re commit 'pass destination dir to unpack rocks archive':

* I don't get why this is needed in the first place. Nothing in the commit
  message helps to understand it (initially my thought was that it adds an
  option to specify a directory to unpack a rock to).
* `source_dir` is passed to `unpack_archive()` in `elseif` branch in
  `fetch_and_change_to_source_dir()`, but here it is always `nil`. Looks as
  the mistake.
* The patch changes several calls to `unpack_archive()`. I tried only one case
  (described in the bullet below). The commit message does not give me any
  pointers what was changed aside of the case below. So to review the patch I
  need to dig into the code and elaborate all possible cases where a behaviour
  can be changed by the patch. I think it is the responsibility of a developer
  to describe what is the behaviour that was changed, how exactly it was
  changed and, more importantly, why.
* I have cloned repo of avro-schema and packed it with `luarocks pack
  avro-schema-scm-1.rockspec` to at least see what the patch doing. Before the
  patch `tarantoolctl rocks unpack /path/to/avro-schema-scm-1.src.rock`
  produces `avro-schema-scm-1/avro-schema` directory with the repository
  content (including the rockspec). After the patch it produces the same
  `avro-schema-scm-1/avro-schema` directory, but it contains only two files:
  `avro-schema` directory with the repository content (including the rockspec)
  and the rockspec (again).
* I see that it is some layout change, but I don't know why this is needed and
  cannot review it. Removed the patch from the patchset.

Re commit 'wrap call other loaders in pcall':

* Again, nothing says why the change is needed.
* A call of a loader was wrapped into a table to handle multireturn. After the
  change only one return value will be returned (a function). While it seems
  that a loader normally returns only one value (a function, a string or nil)
  and is called with lua_call(L, 1, 1) in lj_cf_package_require(), there is no
  reason to discard multireturn here: maybe luarocks itself or, say, busted
  uses expanded convention here (I don't know).
* The [contract][1] of a loader obligates it to return nil or a string in case
  of an error. It should not raise an error.
* So I need to know what is the exact problem the commit solves. Removed the
  patch from the patchset for now.

[1]: https://pgl.yoyo.org/luai/i/package.loaders

Re commit 'stop scanning rocks servers if exact match found':

* Removed trailings spaces removing and empty line changes. Yep, it is bad
  when you leave it in your code, but it is not enough reason to change it
  when it is not a part of your changes. This can make reading your patch,
  blaming changes in the future and backporting patches from an upstream more
  complex. The general rule here is to avoid changes that are not related to
  your patch.
* Removed forgotten commented print (BTW, I already noticed it in [here][2]).
* I see the difference with the original Yaroslav's patches ('Stop scanning
  rocks servers if exact match found' and 'Optimize repos scanning'): you call
  `search.local_manifest_search()` if a file URI is passed as a repository
  URI. Is it the fix for some luarocks regression (if so it should be
  mentioned in the commit message)? I put a rock file for avro-schema into a
  directory, created a repo with `luarocks-admin make_manifest .` and added it
  as the first repository into hardcoded.lua. Then disabled a network and
  tried with `search.local_manifest_search()` and with
  `remote_manifest_search()`. In both cases `tarantoolctl rocks install
  avro-schema scm-1` does not report a network error. So I removed this change
  from the patch.
* Splitted the long condition in `if` as it was made in the 2nd Yaroslav's
  patch.
* Added a link to the problem description to the commit message.
* Mentioned the original commits and the author in the commit message.
* Titlecased the commit header.

[2]: https://github.com/tarantool/tarantool/pull/4207#issuecomment-500205087

I'll continue tomorrow. The plan is the following:

* Review remaining patches in the patchset.
* Verify that patches that are not backported from `tarantool-1.7` branch are
  not needed anymore (in other words that moving to `luarocks-3.1.1-tarantool`
  branch will not lead to a regression).
* Run some commands to ensure that at least basic commands works as expected.
* Verify new commands: at least `tarantoolctl rocks test`.

**Comment by Totktonada** [comment]
Updated the PR again.

Re commit 'Stop scanning rocks servers if exact match found' (again):

* Set the original author: there are no significant changes; removed author
  mention from the commit message.

Re commit 'add support for non-standard path':

* Set the original author for the commit.
* Clarified the problem in the commit message.
* Linked the issue with the problem description.
* Mentioned the original patch in the commit message.
* Titlecased the commit header.
* Moved before 'Stop scanning rocks servers if exact match found', because
  that was the original order in `tarantool-1.7` branch.

Re commit 'Use digest.md5_hex to compute md5 digests instead of openssl':

* Moved right after 'Stop scanning rocks servers if exact match found' as in
  the original patchset (`tarantool-1.7`).
* Set the original commit author.
* Added 'cherry-picked from' clause.

Re commit 'don't check lua when process dependencies':

* Rewritten it in less intrusive way, see the commit 'Skip Lua dynamic library
  search for tarantool'. It only skips the check for a Lua library existence
  when tarantool is set as the Lua interpreter.
* Added proper motivation into the commit message.

Re commit 'test: fix reporting failures on 'command' backend':

* Cherry-picked the version of the patch that was land to the upstream
  (instead of intermediate version).

Re commit 'fix check default lua_version':

* `nil or smth` will always give `smth`.
* It was to give a warnings that Yaroslav removed later in 'Get rid of
  useless warning' commit instead of raising the error re concatenation with
  `nil` (cited below). I performed many tests to understand why your patch is
  needed and only found it by occasion when experimenting around the
  Yaroslav's patch.
* It is trackerized in https://github.com/luarocks/luarocks/issues/961 . Worth
  to mention.
* As I see from the commit where `get_lua_version()` was introduced (a21fcdc4)
  the value it returns is about --lua-ver[sion] and --lua-dir command-line
  options and a saved configuration value (inside an installed rock?). It is
  intended to be `nil` when unset. Different Lua interpreter detection logic
  works depending on whether it is `nil` or not.
* So I doubt this is the right way to fix the issue. I removed the patch from
  the patchset, but fixed the warning to report `cfg.lua_version` when the
  options mentioned above are not specified (commit 'Fix concatenate with a
  nil value when give a warn').

> /usr/share/tarantool/luarocks/cmd.lua:392: attempt to concatenate local
> 'lua_version' (a nil value)'

Re commit 'fix strange string.match incorrect behavior':

* Again, there is no description why this change it needed. I already
  mentioned that in the [PR][3] to tarantool re luarocks-3).
* This function is called from two places and the set of values for which
  :match() is called is known: 'string', 'string?', 'boolean', 'boolean?'. You
  changed '^(.-)(%??)$' to '^(.-)(%?-)$', but they are equivalent when only
  zero or one question mark is at end of a string. I verified this with
  tarantool.
* It seems this patch actually does not change any behaviour. I removed it
  from the patchset.

[3]: https://github.com/tarantool/tarantool/pull/4207#issuecomment-500205087

Re commit 'Get rid of useless warning':

* We can support searching for tarantool as the Lua interpreter instead for
  silence the warning. I think this is good step toward support tarantool in
  the upstream luarocks. Aside of that this way looks more clean for me. I
  made the patch 'Detect tarantool as the Lua interpreter'.
* Now the warnings means for us that hardcoded.lua contains broken paths, so I
  removed your patch about removing this warning from the patchset.

I verified https://github.com/tarantool/tarantool/issues/2609 , found the
regression and cherry-picked 'Add --recursive to git clone' commit from
`tarantool-1.7`.

I think that less intrusive and properly described changes allow us to
upstream most of them. It can help us to reduce maintenance costs in the
future. However I made changes less intrusive only where it was easy to do.
Upstreaming will need more work on that.

I'll continue today with results of testing of the branch.

**Comment by Totktonada** [comment]
I installed the updated version of tarantool into the system (from ebuild on
Gentool Linux) and performed the following tests in an empty directory:

* `tarantoolctl rocks search [substring-of-pkg-name]`
* `tarantoolctl rocks help {<empty>|help|install}`
* install / remove avro-schema, vshard, mysql with and without a version
  specified; run tarantool and require installed modules
* `tarantoolctl rocks list [substring-of-pkg-name]`
* `tarantoolctl rocks show avro-schema`
* `tarantoolctl rocks doc avro-schema`
* `tarantoolctl rocks make_manifest` in a directory with .rocks directory
* `tarantoolctl rocks pack avro-schema` creates
  `avro-schema-scm-1.linux-x86_64.rock` file.
* `tarantoolctl rocks unpack avro-schema-scm-1.linux-x86_64.rock` creates
  `avro-schema-scm-1` directory with the following content:

```
avro-schema-scm-1/
├── avro-schema-scm-1.rockspec
├── doc
│   ├── CHANGELOG.md
│   └── README.md
├── lib
│   └── avro_schema_rt_c.so
├── lua
│   └── avro_schema
│       ├── backend.lua
│       ├── compiler.lua
│       ├── fingerprint.lua
│       ├── frontend.lua
│       ├── il.lua
│       ├── init.lua
│       ├── runtime.lua
│       └── utils.lua
└── rock_manifest
```

I don't know how to verify whether it is good or bad.

* `tarantoolctl rocks unpack vshard` creates `vshard-scm-1/vshard` directory
  with content of the repository. The layout is different from unpack of a
  compiled rock file.
* Installed avro-schema, packed it (compiled rock), removed avro-schema, then
  `tarantoolctl rocks install avro-schema-scm-1.linux-x86_64.rock`.

Everything looks good except the following notes:

* `tarantoolctl rocks help` shows options which are not supported for now. We
  however should support passing options to rocks subcommands and I propose to
  do so in the scope of [#3632][gh-3632].
* `tarantoolctl rocks doc avro-schema` opens README.md (saved locally) in a
  browser. Firefox doesn't want to open it directly, but rather ask me to
  download it. We possibly should provide another way to open a document.
* I cannot verify results of unpack command, because don't know how should a
  layout look.

[gh-3632]: https://github.com/tarantool/tarantool/issues/3632

Then I step into the directory with clone avro-schema module and made the
following tests:

* Slightly changed the module and typed `tarantoolctl rocks make`; verified
  that the changed module is installed into the current directory (to .rocks).
* `tarantoolctl rocks pack avro-schema-scm-1.rockspec` creates
  `avro-schema-scm-1.src.rock` file.
* `tarantoolctl rocks unpack avro-schema-scm-1.src.rock` creates
  `avro-schema-scm-1` directory with the following content:

```
avro-schema-scm-1
├── avro-schema
│   └─ <repository content>
└── avro-schema-scm-1.rockspec
```

* `tarantoolctl rocks install avro-schema-scm-1.src.rock` installs the module
  into .rocks.
* Applied the following diff on avro-schema repo:

```diff
diff --git a/avro-schema-scm-1.rockspec b/avro-schema-scm-1.rockspec
index ddb020b..c995d8c 100644
--- a/avro-schema-scm-1.rockspec
+++ b/avro-schema-scm-1.rockspec
@@ -1,3 +1,4 @@
+rockspec_format = '3.0'
 package = 'avro-schema'
 version = 'scm-1'
 source  = {
@@ -26,4 +27,8 @@ build = {
         TARANTOOL_INSTALL_LUADIR="$(LUADIR)",
     },
 }
+test = {
+    type = 'command',
+    command = 'cmake . && make && make check',
+}
 -- vim: syntax=lua
```

Run `tarantoolctl rocks test`, tests are passed: the exit code is 0. Broke one
test, run again: the exit code is 1.

Everything looks good except maybe:

* I don't know whether the layout produced by unpack is right or not.

To sum up: in my installation the patchset seems to work as expected.

**Comment by Totktonada** [comment]
As Yarosval observed, the problem that is workarounded by the 'Fix concatenate with a nil value when give a warn' patch is already fixed in upstream ( https://github.com/luarocks/luarocks/commit/daf7e7e5f5ede8c3a5a00e2edaf8246209712781 ) in more general way. We discussed it and decided to proceed now with our workaround and remove it when we'll prepare update to the new luarocks version.

**Comment by Totktonada** [comment]
Re 'strange string.match behaviour':

We debugged it with Yaroslav. It was due to interpreting `??)` as the trigraph for `]` ( see https://gcc.gnu.org/onlinedocs/cpp/Initial-processing.html ) when luarocks is bundled into tarantool. We'll fix it in tarantool with escaping of question marks in `extra/txt2c.c`.

**Comment by rosik** [comment]
I've run some tests on static layout (tarantool is built statically, installed in home directory).

1. `Warning: Could not find a Lua interpreter for version 5.1` is still issued because tarantool isn't located in hardcoded.PREFIX. I guess it's not a blocker and we can skip it later.

2. The problem, which was solved by 'Fix strange string.match incorrect behavior' (https://github.com/tarantool/luarocks/commit/f4e2b312a327a68eabc79c51977be0ce098b8cad) was caused by built-in luarocks.
In our environment it was converted to a C file with static string, which contained a trigraph `??)`/ 

GCC issued a warning:
```
/tarantool/src/../third_party/luarocks/src/luarocks/util.lua.c:715:61: warning:
                                                 trigraph ??) converted to ] [-Wtrigraphs]
 "         local vo, optional = valid_opts[k]:match(\"^(.-)(%??)$\")\n"
                                                             ^
```
We decided to fix it later here: https://github.com/tarantool/tarantool/blob/master/extra/txt2c.c

3. I've tested commands `make`, `install`, `doc`, `list`, `show`, `pack`, `search` and they seem to work as expected.

**Comment by rosik** [comment]
4. `tarantoolctl rocks test` seems to work too.

**Review by rosik — APPROVED** [review]

**Review by Totktonada — APPROVED** [review]
It seems the patchset handles all needs of the standard tarantool layout (when it is installed from a package manager). Portable layout or using `tarantoolctl rocks` directly from a build directory can lead to a warning re unability to find a Lua interpreter (it is not an error). We will handle it later I think.
